zookeeper-server/src/main/java/org/apache/zookeeper/Login.java

```diff
@@ -286,6 +286,13 @@ public Subject getSubject() {
         return subject;
     }
 
+    public String getUserName() {
+        if (principal == null || principal.isEmpty()) {
+            return System.getProperty("user.name", "<NA>");
```

Comment:
You can sample the value as a static variable. You will save an access to the system properties map

Reply:
done